File: Vigenere/vigenere.c
//Brijesh Patel
// CIS 3360

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

int keyLength;
// takes content from file and stores it in an array after performing padding and punctutation cleaning
char *copyToArray(char *fileName, char *a, int flag)
{
	int i = 0, len, j = 0;
	char c, *temp = NULL;
	FILE *fp = fopen(fileName, "r");

	while(fscanf(fp, " %c", &c) != EOF)
	{
		if(!isalpha(c))
			continue;
		// only store lower case  alpabetic characters
		a[i++] = tolower(c);
	}
	// don't pad the key just return
	if(flag == 1)
		{
			fclose(fp);
			return NULL;
		}

	len = strlen(a);
	// pad with 'x's if neccesary 
	if(len < 512)
	{	
		temp = malloc(sizeof(char) * 512);
		for(i = 0; i < 512; i++)
		{
			if(i >= len)
			{
				//if(i % keyLength == 0)
					//break;
				temp[j++] = 'x';
				continue;
			}
			temp[j++] = a[i];
		}
		fclose(fp);
		return temp;
	}
	// trunctate characters that are beyond the 512 limit
	else if(len > 512)
	{
		temp = malloc(sizeof(char) * 512);

		for(i = 0; i < 512; i++)
			temp[i] = a[i];	

		fclose(fp);
		return temp;
	}
	fclose(fp);
	return a;
	
}
// encrypt plaintext array and output cipher array
char *vigenereCipher(char *p, char *key)
{
	int i, j = 0, len;
	char *temp = malloc(sizeof(char)* 512), tmp, tmp2;

	len = strlen(p);

	for(i = 0; i < len; i++)
	{	
		// restart key from beginning(reusing the key)
		if(j == keyLength)
			j = 0;

		tmp = key[j++] - 97;
		tmp2 = p[i] - 97;

		int val = ((tmp + tmp2) % 26);

		temp[i] = (char) (val + 97);
	}
	temp[i] = '\0';
	return temp;
}


int main(int argc, char **argv)
{
	int i = 0, len, j;
	char *keyFile = argv[1];
	char *plainTextFile= argv[2];
	char *key = malloc(sizeof(char) * 512);
	char *p = malloc(sizeof(char) * 512);
	char *cipherText = malloc(sizeof(char) * 512);

	printf("\n\nVigenere Key:\n\n");

	copyToArray(keyFile, key, 1);

	len  = strlen(key);
	keyLength = len;
	for(i = 0; i < len;)
	{
		for(j = 0; j < 80 && i < len; j++)
		{
			printf("%c", key[i++]);
		}
		printf("\n");
	}

	printf("\n\n");
	printf("Plaintext:\n\n");

	p = copyToArray(plainTextFile, p, 0);
	len = strlen(p);

	for(i = 0; i < len;)
	{
		for(j = 0; j < 80 && i < len; j++)
		{
			printf("%c", p[i++]);
		}
		printf("\n");
	}

	printf("\n\n");
	printf("Ciphertext:\n\n");

	cipherText = vigenereCipher(p, key);
	
	len = strlen(cipherText);
	for(i = 0; i < len;)
	{
		for(j = 0; j < 80 && i < len; j++)
		{	
			// prevents left over characters in the array from being printed
			if(cipherText[i] == '\0')
				break;
			printf("%c", cipherText[i++]);
		}
		printf("\n");
	}

	free(cipherText);
	free(p);
	free(key);

}

File: Vigenere/vigenere.java


import java.io.*;
import java.util.Scanner;
public class vigenere
{
	public static PrintWriter pw;
	public static int keyLen;
	public static int copyToArray(String fileName,char a[], int flag) throws IOException
	{
		File file = new File(fileName);
		String str = "";
		int j = 0, size = 0;
		Scanner scanner = new Scanner(file);
		char temp[] = new char[512];
		while(scanner.hasNext())
		{
			str = scanner.nextLine();
			str.replace("\0", "");
			
			for(int i = 0; i < str.length() && size < 512; i++)
				{	
					if(str.charAt(i) == '^' || str.charAt(i) == '_')
						continue;
					if(isLetter(str.charAt(i)))
					{
					temp[j++] = str.charAt(i); 
					size++;
					}	
				}
			
		}
		int tmp = size;
		size = 0;
		j = 0;
		//convert all characters in array to lowercase
		for(int i = 0; i < tmp; i++)
		{
			a[j++] = Character.toLowerCase(temp[i]);
			size++; 
			
		}
		// Don't pad key just return
		if(flag == 1)
		{
			a = temp;
			scanner.close();
			return size;
		}
		
		
		// need to pad 
		if(tmp < 512)
		{	
			tmp = size;
			size = 0;
			j = 0;
			temp = new char[512];
			for(int i = 0; i < 512;i++)
			{	
				if(i >= tmp)
				{	
					temp[j++] = 'x';
					size++;
				}
				if(i < tmp)
				{
					temp[j++] = a[i];
					size++;
				}
			 }
			j = 0;
			for(int i = 0; i < size; i++)
			{
				a[j++] = temp[i];
			}
			if(size < 512)
				a[size] = '\0';
		}
		
		scanner.close();
		return size;
	}
	// check for alphabetic characters only 
	private static boolean isLetter(char c) {
		
		if(c < 65 || c > 122)
			return false;
		
		return true;
	}

	public static void main(String args[]) throws IOException
	{
			
		String keyFile = args[0];
		String plainTextFile = args[1];
		char keyText[] = new char[512];
		char plainText[] = new char[512];
		char cipherText[] = new char[512];
		int keySize, plainSize, j = 0;

		pw = new PrintWriter(System.out);
		
		pw.printf("\n\nVigenere Key:\n\n");
		pw.flush();
		keySize = copyToArray(keyFile, keyText, 1);
		keyLen = keySize;
		//print key
		for(int i = 0; i < keySize;)
		{
			for(j = 0; j < 80 && i < keySize; j++)
			{
				pw.printf("%c", keyText[i++]);
				pw.flush();
			}
			
			System.out.println();
			
		}

		pw.printf("\n\nPlaintext:\n\n");
		pw.flush();
		
		// next capture the plain text
		plainSize = copyToArray(plainTextFile,plainText, 0);
		// print plain text
		for(int i = 0; i < plainSize;)
		{
			for(j = 0; j < 80 && i < plainSize;j++)
			{
				if(plainText[i] == '\0')
					break;
				pw.printf("%c", plainText[i++]);
				pw.flush();
			}
			pw.println();
		}

		pw.printf("\n\nCiphertext:\n\n");
		pw.flush();
		
		cipherText = vigenere(plainText, keyText, plainSize);
		//print ciphertext
		for(int i = 0; i < plainSize;)
		{
			for(j = 0; j < 80 && i < plainSize;j++)
			{
				if(plainText[i] == '\0')
					break;
				pw.printf("%c", cipherText[i++]);
				pw.flush();
			}
			pw.println();
		}
		System.out.println();
	
	}
	
	private static char[] vigenere(char []plain, char[]key, int len)
	{

			int i, j = 0;
			char []temp = new char[512];
			int tmp, tmp2;
			
			for(i = 0; i < len; i++)
			{	
				// restart key from beginning(reusing the key)
				if(j == keyLen)
					j = 0;
			
				tmp = key[j++] - 97;
				tmp2 = plain[i] - 97;
				
				int val = ((tmp + tmp2) % 26);

				temp[i] = (char) (val + 97);
			
			}
			if(i != 512)
				temp[i] = '\0';
			return temp;
	}

}

